### PR TITLE
feat(setup-checklist): add configure SSO notification alert

### PR DESCRIPTION
**Pull request by erzhtor**
Closes https://github.com/sourcegraph/sourcegraph/issues/55433.

This PR adds a warning notification/alert if an enterprise license and no SSO providers are configured.

## Test plan
- Added new tests

## Screenshots
image

### CODE REVIEW

File: cmd/frontend/graphqlbackend/site_alerts_test.go [hunk 1]
```diff
@@ -113,34 +113,27 @@ func TestDetermineOutOfDateAlert(t *testing.T) {
 
 func TestObservabilityActiveAlertsAlert(t *testing.T) {
 	f := false
-	type args struct {
-		args AlertFuncArgs
-	}
 	tests := []struct {
 		name string
-		args args
+		args AlertFuncArgs
```
Comment (sashaostrikov) [inline cmd/frontend/graphqlbackend/site_alerts_test.go:118]: Nice!

File: enterprise/cmd/frontend/internal/auth/init.go [hunk 2]
```diff
@@ -60,60 +61,68 @@ func Init(logger log.Logger, db database.DB) {
...
+			name = "SAML"
+		default:
+			name = "Other"
 		}
-		if len(names) == 0 {
-			return nil
+
+		if _, ok := collected[name]; !ok {
+			collected[name] = struct{}{}
+			names = append(names, name)
```
Comment (BolajiOlajide) [inline enterprise/cmd/frontend/internal/auth/init.go:103]:
```suggestion
		if !collected.Has(name) {
			collected.Add(name)
```

We can make use of the methods available to sets here.

File: enterprise/cmd/frontend/internal/auth/init.go [hunk 3]
```diff
@@ -60,60 +61,68 @@ func Init(logger log.Logger, db database.DB) {
...
 		}
+	}
 
+	if len(names) > 0 && !licensing.IsFeatureEnabledLenient(licensing.FeatureSSO) {
```
Comment (BolajiOlajide) [inline enterprise/cmd/frontend/internal/auth/init.go:107]:
```suggestion
	names := collected.Values()
	if len(names) > 0 && !licensing.IsFeatureEnabledLenient(licensing.FeatureSSO) {
```

This way, we keep track of the names in one data structure instead of two.

File: enterprise/cmd/frontend/internal/auth/init.go [hunk 4]
```diff
@@ -60,60 +61,68 @@ func Init(logger log.Logger, db database.DB) {
...
+
+func requireLicenseOrSuggestSSOAlerts(args graphqlbackend.AlertFuncArgs) []*graphqlbackend.Alert {
+	// Only site admins can act on this alert, so only show it to site admins.
+	if !args.IsSiteAdmin {
+		return nil
+	}
 
-		if licensing.IsFeatureEnabledLenient(licensing.FeatureSSO) {
-			return nil
+	collected := make(map[string]struct{})
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/init.go:73]: nit: you can use `collections.Set` and simplify the code a bit

**Review by BolajiOlajide — APPROVED** [review]

File: enterprise/cmd/frontend/internal/auth/init_test.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+			isSiteAdmin:        true,
+			want: []*graphqlbackend.Alert{{
+				GroupValue:                graphqlbackend.AlertGroupAuthentication,
+				TypeValue:                 graphqlbackend.AlertTypeWarning,
+				MessageValue:              "We recommend that enterprise instances use SSO or SAML to authenticate users. [Set up authentication now](/site-admin/configuration)",
+				IsDismissibleWithKeyValue: "configure-sso-providers",
+			}},
+		},
+		{
+			name:               "do not show alert if no SSO providers configured and license has SSO feature and and feature flag is disabled",
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/init_test.go:65]:
```suggestion
			name:               "do not show alert if no SSO providers configured and license has SSO feature and feature flag is disabled",
```

File: enterprise/cmd/frontend/internal/auth/init.go [hunk 6]
```diff
@@ -60,60 +61,68 @@ func Init(logger log.Logger, db database.DB) {
...
 		sort.Strings(names)
 		return []*graphqlbackend.Alert{{
 			GroupValue:   graphqlbackend.AlertGroupLicense,
 			TypeValue:    graphqlbackend.AlertTypeError,
 			MessageValue: fmt.Sprintf("A Sourcegraph license is required to enable following authentication providers: %s. [**Get a license.**](/site-admin/license)", strings.Join(names, ", ")),
 		}}
-	})
+	}
+
+	if len(names) == 0 && licensing.IsFeatureEnabledLenient(licensing.FeatureSSO) && featureflag.FromContext(args.Ctx).GetBoolOr("setup-checklist", false) {
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/init.go:116]: Q: Why don't we want to show the SSO warning if setup checklist is disabled?

File: enterprise/cmd/frontend/internal/auth/init_test.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+			ctx := actor.WithActor(context.Background(), actor.FromMockUser(1))
+			if test.featureFlagEnabled {
+				ctx = featureflag.WithFlags(ctx, featureflag.NewMemoryStore(map[string]bool{"setup-checklist": true}, nil, nil))
+			}
+			// run
+			gotAlerts := requireLicenseOrSuggestSSOAlerts(graphqlbackend.AlertFuncArgs{
+				IsSiteAdmin: test.isSiteAdmin,
+				Ctx:         ctx,
+			})
+			// checkks
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/init_test.go:121]:
```suggestion
			// checks
```

File: enterprise/cmd/frontend/internal/auth/init_test.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+				return licensing.NewFeatureNotActivatedError("test")
+			}
+			t.Cleanup(func() {
+				conf.Mock(nil)
+				licensing.MockCheckFeature = oldMock
+			})
+			ctx := actor.WithActor(context.Background(), actor.FromMockUser(1))
+			if test.featureFlagEnabled {
+				ctx = featureflag.WithFlags(ctx, featureflag.NewMemoryStore(map[string]bool{"setup-checklist": true}, nil, nil))
+			}
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/init_test.go:115]: It might be worth extracting the setup into a separate function receiving a test case. This way the test body will read better and be more concise.

**Review by sashaostrikov — COMMENTED** [review]
Sidenote not directly related to this exact PR:
Do we need to have both a dot and a number on this icon? Maybe just a number is enough?
CleanShot 2023-08-16 at 3 31 10

File: enterprise/cmd/frontend/internal/auth/init.go (hunk 4 above)
Comment (kopancek) [reply enterprise/cmd/frontend/internal/auth/init.go:73]: Thanks Alex for promoting the best collection :)

File: enterprise/cmd/frontend/internal/auth/init.go (hunk 6 above)
Comment (kopancek) [reply enterprise/cmd/frontend/internal/auth/init.go:116]: All of the code that we've been working on so far is behind a feature flag. We'll remove it once we are happy with the result.

**Review by kopancek — APPROVED** [review]
Looks good to me, but agree with all the other comments left before. Once those are addressed, we can merge. Approving to unblock merging once you're done.

File: enterprise/cmd/frontend/internal/auth/init.go (hunk 2 above)
Comment (erzhtor) [reply enterprise/cmd/frontend/internal/auth/init.go]: Thank you, updated a6ca4bcf1ae29b77d677e8a226c43260c3de24a8.

File: enterprise/cmd/frontend/internal/auth/init.go (hunk 3 above)
Comment (erzhtor) [reply enterprise/cmd/frontend/internal/auth/init.go:105]: Thank you, updated a6ca4bcf1ae29b77d677e8a226c43260c3de24a8.

File: enterprise/cmd/frontend/internal/auth/init.go (hunk 4 above)
Comment (erzhtor) [reply enterprise/cmd/frontend/internal/auth/init.go]: Thank you, updated a6ca4bcf1ae29b77d677e8a226c43260c3de24a8.

**Comment by erzhtor** [comment]
> Sidenote not directly related to this exact PR: Do we need to have both a dot and a number on this icon? Maybe just a number is enough? CleanShot 2023-08-16 at 3 31 10

That is a good point. We will change the UI once we have finalized the designs. For now, this is just what we proactively devised to unblock ourselves.

File: enterprise/cmd/frontend/internal/auth/init.go (hunk 6 above)
Comment (sashaostrikov) [reply enterprise/cmd/frontend/internal/auth/init.go:114]: Gotcha. But I wonder whether the SSO warning is not related only to the current project and we might benefit from this warning even right now?

File: enterprise/cmd/frontend/internal/auth/init.go [hunk 9]
```diff
@@ -60,60 +61,66 @@ func Init(logger log.Logger, db database.DB) {
...
+			name = "SAML"
+		default:
+			name = "Other"
 		}
-		if len(names) == 0 {
-			return nil
+
+		if !names.Has(name) {
+			names.Add(name)
 		}
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/auth/init.go:102]: Nitpick drive-by: it's a set, why check whether it has the `name`? :)

**Comment by unknwon** [comment]
Nice 🌟 

Though I have to (repeatedly) point out that "SSO" and "SAML" are not equivalent terms should be put in an AND relation, because SAML is an example of SSO 😂
